core/src/test/java/org/apache/iceberg/TestBaseIncrementalAppendScan.java

```diff
@@ -0,0 +1,121 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestBaseIncrementalAppendScan extends ScanTestBase<IncrementalAppendScan> {
+  public TestBaseIncrementalAppendScan(int formatVersion) {
+    super(formatVersion);
+  }
+
+  @Override
+  protected IncrementalAppendScan newScan() {
+    return table.newIncrementalAppendScan();
+  }
+
+  @Test
+  public void testFromSnapshotInclusive() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    Snapshot snapshotA = table.currentSnapshot();
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Snapshot snapshotB = table.currentSnapshot();
+    table.newFastAppend().appendFile(FILE_C).commit();
+    Snapshot snapshotC = table.currentSnapshot();
+
+    IncrementalAppendScan scan = newScan()
+        .fromSnapshotInclusive(snapshotA.snapshotId());
+    Assert.assertEquals(3, Iterables.size(scan.planFiles()));
+  }
+
+  @Test
+  public void testFromSnapshotExclusiveAndToSnapshot() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    Snapshot snapshotA = table.currentSnapshot();
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Snapshot snapshotB = table.currentSnapshot();
+    table.newFastAppend().appendFile(FILE_C).commit();
+    Snapshot snapshotC = table.currentSnapshot();
+
+    IncrementalAppendScan scan = newScan()
+        .fromSnapshotExclusive(snapshotA.snapshotId())
+        .toSnapshot(snapshotB.snapshotId());
+    Assert.assertEquals(1, Iterables.size(scan.planFiles()));
+  }
+
+  @Test
+  public void testToSnapshot() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    Snapshot snapshotA = table.currentSnapshot();
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Snapshot snapshotB = table.currentSnapshot();
+    table.newFastAppend().appendFile(FILE_C).commit();
+    Snapshot snapshotC = table.currentSnapshot();
+
+    IncrementalAppendScan scan = newScan()
+        .toSnapshot(snapshotB.snapshotId());
+    Assert.assertEquals(2, Iterables.size(scan.planFiles()));
+  }
+
+  @Test
+  public void testMultipleRootSnapshots() throws Exception {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    Snapshot snapshotA = table.currentSnapshot();
+
+    table.manageSnapshots()
+        .createBranch("branch", snapshotA.snapshotId())
+        .commit();
+
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Snapshot snapshotB = table.currentSnapshot();
+    table.newFastAppend().appendFile(FILE_C).commit();
+    Snapshot snapshotC = table.currentSnapshot();
+
+    // sleep a milli second so that snapshotD will certainly have a different timestamp
+    Thread.sleep(1);
```

Comment:
We typically use `waitUntilAfter(snapshotC.timestampMillis())` to avoid `Thread.sleep`. It does basically the same thing, but ensures that you have a different timestamp.